drv/lpc55-flash/src/lib.rs

```diff
@@ -0,0 +1,370 @@
+//! A raw driver for the LPC55 flash controller without using the ROM.
+//!
+//! This driver is written in a very generic form that doesn't assume any
+//! particular execution model. You can use it to implement busy waiting, or
+//! an interrupt driven driver, etc.
+//!
+//! See the [`Flash`] type for more details.
+
+#![no_std]
+
+use core::ops::RangeInclusive;
+
+/// Number of bytes per flash word.
+pub const BYTES_PER_FLASH_WORD: usize = 16;
+
+/// Number of bytes per flash page.
+pub const BYTES_PER_FLASH_PAGE: usize = 512;
+
+/// Flash driver handle. Wraps a pointer to the flash register set and provides
+/// encapsulated operations.
+///
+/// # Procedure for writing to flash
+///
+/// 1. Use `start_erase_region` to begin an erase of your desired region. Use
+///    `poll_erase_or_program_result` to discover when it's done. (Note: you can
+///    use `start_blank_check` to skip this step if there's a chance the flash
+///    is already erased.)
+/// 2. Call `start_write_row` with each 16 byte chunk of the first page you're
+///    writing, 32 times, for rows 0-31. Each time, use `poll_write_result` to
+///    check that it's done.
+/// 3. Call `start_program` with the index of any word in the target flash page
+///    to begin the programming process. Use `poll_erase_or_program_result` to
+///    learn when it's done.
+/// 4. Repeat steps 2 and 3 for each page in the region you want to write.
+///
+/// # Flash addressing and word numbers
+///
+/// While the flash is directly addressable by the processor at byte
+/// granularity, most of the flash controller operations work in terms of a
+/// _word number._ Internal flash words are 16 bytes (128 bits) in length; the
+/// word number is the index of a word within the flash, starting at 0.
+///
+/// Converting an address to a word number:
+///
+/// - Divide by 16 (or right-shift by 4).
+/// - Mask all but the bottom 18 bits.
```

Comment:
If you're taking an address that points into the direct-mapped flash space, sure! Access to the Flash secure alias is the easiest case.

The shift-and-mask in these two lines will basically hide any data above the 22nd bit (because it's right-shifted by 4 and then masked). That's `0x40_0000` and up. Accessing the secure alias will additionally set bit 28, but that bit doesn't matter from the perspective of the flash controller.